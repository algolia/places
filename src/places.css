.algolia-places {
  width: 100%;
}

.ap-input, .ap-hint {
  width: 100%;
  padding-right: 35px;
  padding-left: 16px;
  line-height: 40px;
  height: 40px;
  border: 1px solid #CCC;
  border-radius: 3px;
  outline: none;
  font: inherit;
  appearance: none;
  -webkit-appearance: none;
  box-sizing: border-box;
}

.ap-input::-webkit-search-decoration {
  -webkit-appearance: none;
}

.ap-input::-ms-clear {
  display: none;
}

.ap-input:hover ~ .ap-input-icon svg,
.ap-input:focus ~ .ap-input-icon svg,
.ap-input-icon:hover svg {
  fill: #aaaaaa;
}

.ap-dropdown-menu {
  width: 100%;
  background: #ffffff;
  box-shadow: 0 1px 10px rgba(0, 0, 0, 0.2), 0 2px 4px 0 rgba(0, 0, 0, 0.1);
  border-radius: 3px;
  margin-top: 3px;
  overflow: hidden;
}

.ap-suggestion {
  cursor: pointer;
  height: 46px;
  line-height: 46px;
  padding-left: 18px;
  overflow: hidden;
}

.ap-suggestion em {
  font-weight: bold;
  font-style: normal;
}

.ap-address {
  font-size: smaller;
  margin-left: 12px;
  color: #aaaaaa;
}

.ap-suggestion-icon {
  margin-right: 10px;
  width: 14px;
  height: 20px;
  vertical-align: middle;
}

.ap-suggestion-icon svg {
  -webkit-transform: scale(0.9) translateY(2px);
          transform: scale(0.9) translateY(2px);
  fill: #cfcfcf;
}

.ap-input-icon {
  border: 0;
  background: transparent;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 16px;
  outline: none;
}

.ap-input-icon.ap-icon-pin {
  cursor: initial;
}

.ap-input-icon svg {
  fill: #cfcfcf;
  position: absolute;
  top: 50%;
  right: 0;
  -webkit-transform: translateY(-50%);
          transform: translateY(-50%);
}

.ap-cursor {
  background: #efefef;
}

.ap-cursor .ap-suggestion-icon svg {
  -webkit-transform: scale(1) translateY(2px);
          transform: scale(1) translateY(2px);
  fill: #aaaaaa;
}

.ap-footer {
  opacity: .8;
  text-align: right;
  padding: .5em 1em .5em 0;
  font-size: 12px;
  line-height: 12px;
}

.ap-footer a {
  color: inherit;
  text-decoration: none;
}

.ap-footer a svg {
  vertical-align: text-bottom;
  max-width: 60px;
}

.ap-footer:hover {
  opacity: 1;
}
