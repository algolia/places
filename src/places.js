import EventEmitter from 'events';

import algoliasearch from 'algoliasearch/src/browser/builds/algoliasearchLite';
import autocomplete from 'autocomplete.js';

import './navigatorLanguage';

import createAutocompleteDataset from './createAutocompleteDataset';

import clearIcon from './icons/clear.svg';
import pinIcon from './icons/address.svg';

import css from './places.css';
import insertCss from 'insert-css';
insertCss(css, { prepend: true });

import errors from './errors';
import createReverseGeocodingSource from './createReverseGeocodingSource';

const applyAttributes = (elt, attrs) => {
  Object.entries(attrs).forEach(([name, value]) => {
    elt.setAttribute(name, `${value}`);
  });

  return elt;
};

export default function places(options) {
  const {
    container,
    style,
    accessibility,
    autocompleteOptions: userAutocompleteOptions = {},
  } = options;

  // multiple DOM elements targeted
  if (container instanceof NodeList) {
    if (container.length > 1) {
      throw new Error(errors.multiContainers);
    }

    // if single node NodeList received, resolve to the first one
    return places({ ...options, container: container[0] });
  }

  // container sent as a string, resolve it for multiple DOM elements issue
  if (typeof container === 'string') {
    const resolvedContainer = document.querySelectorAll(container);
    return places({ ...options, container: resolvedContainer });
  }

  // if not an <input>, error
  if (!(container instanceof HTMLInputElement)) {
    throw new Error(errors.badContainer);
  }

  const placesInstance = new EventEmitter();
  const prefix = `ap${style === false ? '-nostyle' : ''}`;

  const autocompleteOptions = {
    autoselect: true,
    hint: false,
    cssClasses: {
      root: `algolia-places${style === false ? '-nostyle' : ''}`,
      prefix,
    },
    debug: process.env.NODE_ENV === 'development',
    ...userAutocompleteOptions,
  };

  const autocompleteDataset = createAutocompleteDataset({
    ...options,
    algoliasearch,
    onHits: ({ hits, rawAnswer, query }) =>
      placesInstance.emit('suggestions', {
        rawAnswer,
        query,
        suggestions: hits,
      }),
    onError: (e) => placesInstance.emit('error', e),
    onRateLimitReached: () => {
      const listeners = placesInstance.listenerCount('limit');
      if (listeners === 0) {
        console.log(errors.rateLimitReached); // eslint-disable-line no-console
        return;
      }

      placesInstance.emit('limit', { message: errors.rateLimitReached });
    },
    onInvalidCredentials: () => {
      if (options && options.appId && options.appId.startsWith('pl')) {
        console.error(errors.invalidCredentials); // eslint-disable-line no-console
      } else {
        console.error(errors.invalidAppId); // eslint-disable-line no-console
      }
    },
    container: undefined,
  });

  const autocompleteInstance = autocomplete(
    container,
    autocompleteOptions,
    autocompleteDataset
  );
  const autocompleteContainer = container.parentNode;

  const autocompleteChangeEvents = ['selected', 'autocompleted'];

  autocompleteChangeEvents.forEach((eventName) => {
    autocompleteInstance.on(`autocomplete:${eventName}`, (_, suggestion) => {
      placesInstance.emit('change', {
        rawAnswer: suggestion.rawAnswer,
        query: suggestion.query,
        suggestion,
        suggestionIndex: suggestion.hitIndex,
      });
    });
  });
  autocompleteInstance.on('autocomplete:cursorchanged', (_, suggestion) => {
    placesInstance.emit('cursorchanged', {
      rawAnswer: suggestion.rawAnswer,
      query: suggestion.query,
      suggestion,
      suggestionIndex: suggestion.hitIndex,
    });
  });

  const clear = document.createElement('button');
  clear.setAttribute('type', 'button');
  clear.setAttribute('aria-label', 'clear');
  if (
    accessibility &&
    accessibility.clearButton &&
    accessibility.clearButton instanceof Object
  ) {
    applyAttributes(clear, accessibility.clearButton);
  }
  clear.classList.add(`${prefix}-input-icon`);
  clear.classList.add(`${prefix}-icon-clear`);
  clear.innerHTML = clearIcon;
  autocompleteContainer.appendChild(clear);
  clear.style.display = 'none';

  const pin = document.createElement('button');
  pin.setAttribute('type', 'button');
  pin.setAttribute('aria-label', 'focus');
  if (
    accessibility &&
    accessibility.pinButton &&
    accessibility.pinButton instanceof Object
  ) {
    applyAttributes(pin, accessibility.pinButton);
  }
  pin.classList.add(`${prefix}-input-icon`);
  pin.classList.add(`${prefix}-icon-pin`);
  pin.innerHTML = pinIcon;
  autocompleteContainer.appendChild(pin);

  pin.addEventListener('click', () => {
    autocompleteDataset.source.configure({ useDeviceLocation: true });
    autocompleteInstance.focus();
    placesInstance.emit('locate');
  });

  clear.addEventListener('click', () => {
    autocompleteInstance.autocomplete.setVal('');
    autocompleteInstance.focus();
    clear.style.display = 'none';
    pin.style.display = '';
    placesInstance.emit('clear');
  });

  let previousQuery = '';

  const inputListener = () => {
    const query = autocompleteInstance.val();
    if (query === '') {
      pin.style.display = '';
      clear.style.display = 'none';
      if (previousQuery !== query) {
        placesInstance.emit('clear');
      }
    } else {
      clear.style.display = '';
      pin.style.display = 'none';
    }
    previousQuery = query;
  };

  autocompleteContainer
    .querySelector(`.${prefix}-input`)
    .addEventListener('input', inputListener);

  const autocompleteIsomorphicMethods = ['open', 'close'];
  autocompleteIsomorphicMethods.forEach((methodName) => {
    placesInstance[methodName] = (...args) => {
      autocompleteInstance.autocomplete[methodName](...args);
    };
  });

  placesInstance.getVal = () => {
    return autocompleteInstance.val();
  };

  placesInstance.destroy = (...args) => {
    autocompleteContainer
      .querySelector(`.${prefix}-input`)
      .removeEventListener('input', inputListener);

    autocompleteInstance.autocomplete.destroy(...args);
  };

  placesInstance.setVal = (...args) => {
    previousQuery = args[0];
    if (previousQuery === '') {
      pin.style.display = '';
      clear.style.display = 'none';
    } else {
      clear.style.display = '';
      pin.style.display = 'none';
    }
    autocompleteInstance.autocomplete.setVal(...args);
  };

  placesInstance.autocomplete = autocompleteInstance;

  placesInstance.search = (query = '') =>
    new Promise((resolve) => {
      autocompleteDataset.source(query, resolve);
    });

  placesInstance.configure = (configuration) => {
    const safeConfig = { ...configuration };

    delete safeConfig.onHits;
    delete safeConfig.onError;
    delete safeConfig.onRateLimitReached;
    delete safeConfig.onInvalidCredentials;
    delete safeConfig.templates;

    autocompleteDataset.source.configure(safeConfig);
    return placesInstance;
  };

  placesInstance.reverse = createReverseGeocodingSource({
    ...options,
    algoliasearch,
    formatInputValue: (options.templates || {}).value,
    onHits: ({ hits, rawAnswer, query }) =>
      placesInstance.emit('reverse', {
        rawAnswer,
        query,
        suggestions: hits,
      }),
    onError: (e) => placesInstance.emit('error', e),
    onRateLimitReached: () => {
      const listeners = placesInstance.listenerCount('limit');
      if (listeners === 0) {
        console.log(errors.rateLimitReached); // eslint-disable-line no-console
        return;
      }

      placesInstance.emit('limit', { message: errors.rateLimitReached });
    },
    onInvalidCredentials: () => {
      if (options && options.appId && options.appId.startsWith('pl')) {
        console.error(errors.invalidCredentials); // eslint-disable-line no-console
      } else {
        console.error(errors.invalidAppId); // eslint-disable-line no-console
      }
    },
  });

  return placesInstance;
}
