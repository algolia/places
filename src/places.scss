@mixin placeholder {
  $placeholders: ":-webkit-input" ":-moz" "-moz" "-ms-input";
  @each $placeholder in $placeholders {
    &:#{$placeholder}-placeholder {
      @content;
    }
  }
}

// all em units are computed given a font of 16px
// see http://pxtoem.com/ to change/add values
.algolia-places {
  width: 100%;
}

.ap-input, .ap-hint {
  width: 100%;
  padding-right: 35px;
}

.ap-input:hover ~ .ap-input-icon svg,
.ap-input:focus ~ .ap-input-icon svg,
.ap-input-icon:hover svg {
  fill: #aaaaaa;
}

// Dropdown
.ap-dropdown-menu {
  width: 100%;
  background: #ffffff;
  box-shadow: 0 1px 10px rgba(0, 0, 0, 0.2), 0 2px 4px 0 rgba(0, 0, 0, 0.1);
  border-radius: 3px;
  margin-top: 3px;
  overflow: hidden; // avoid non rounded borders to overlap
}

.ap-suggestion {
  cursor: pointer;
  height: 46px;
  line-height: 46px;
  padding-left: 18px;
  overflow: hidden;

  em {
    font-weight: bold;
    font-style: normal;
  }
}

.ap-address {
  font-size: smaller;
  margin-left: 12px; // 12px when font is 14px
  color: #aaaaaa;
}

.ap-suggestion-icon {
  margin-right: 10px;
  width: 14px;
  height: 20px;
  vertical-align: middle; // handles right alignment with text

  svg {
    transform: scale(0.90) translateY(2px);
    fill: #cfcfcf;
  }
}

.ap-input-icon {
  border: 0;
  background: transparent;
  position: absolute;
  top: 0;
  bottom: 0;
  right: 16px;
  outline: none;

  svg {
    fill: #cfcfcf;
    // keep the icons centered
    position: absolute;
    top: 50%;
    right: 0;
    transform: translateY(-50%);
  }
}

.ap-cursor {
  background: #efefef;

  .ap-suggestion-icon svg {
    transform: scale(1);
    fill: #aaaaaa;
  }
}

.ap-footer {
  opacity: .8;
  text-align: right;
  padding: .5em 1em .5em 0;
  font-size: 12px;
  line-height: 12px;

  a {
    color: inherit;
    text-decoration: none;
    svg {
      vertical-align: middle;
    }
  }

  &:hover {
    opacity: 1;
  }
}
