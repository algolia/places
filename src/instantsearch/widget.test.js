import searchHelper, { SearchParameters } from 'algoliasearch-helper';
import algoliaPlacesWidget from './widget';
import places from '../places';

jest.mock('../places.js', () => {
  const module = jest.fn(() => {
    const instance = { on: jest.fn(), setVal: jest.fn(), close: jest.fn() };

    module.__instance = instance;
    return instance;
  });

  return module;
});

const createFakeClient = () => ({
  addAlgoliaAgent: () => {},
});

const createFakekHelper = client => {
  const helper = searchHelper(client);

  helper.search = jest.fn();

  return helper;
};

describe('instantsearch widget', () => {
  const defaultOptions = {
    places: 'option',
  };

  it('creates a places instance', () => {
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget(defaultOptions);

    widget.init({ helper });

    expect(places).toHaveBeenCalledWith({ places: 'option' });
  });

  it('creates a places instance without parameters', () => {
    expect(() => algoliaPlacesWidget()).not.toThrow();
  });

  it('configures the helper', () => {
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget(defaultOptions);

    widget.init({ helper });

    expect(helper.getState()).toMatchObject({
      insideBoundingBox: undefined,
      aroundLatLng: undefined,
    });
  });

  it('does not call setQueryParameter during the init step', () => {
    // Using setQueryParameter to change a filter value during the init step will
    // have unintented consequences such as resetting the pagination to 0.
    // We should not do that
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget(defaultOptions);

    helper.setQueryParameter = jest.fn();
    widget.init({ helper });

    expect(helper.setQueryParameter).not.toHaveBeenCalled();
  });

  it('accepts a defaultPosition parameter', () => {
    const widget = algoliaPlacesWidget({ defaultPosition: [1, 1] });

    const beforeConfiguration = {};
    const afterConfiguration = widget.getConfiguration(beforeConfiguration);

    expect(afterConfiguration).toEqual({
      insideBoundingBox: undefined,
      aroundLatLngViaIP: false,
      aroundLatLng: '1,1',
    });
  });

  it('overrides the aroundLatLng if a defaultPosition parameter is passed', () => {
    const widget = algoliaPlacesWidget({ defaultPosition: [1, 1] });

    const beforeConfiguration = {
      aroundLatLng: '12,14',
    };

    const afterConfiguration = widget.getConfiguration(beforeConfiguration);

    expect(afterConfiguration).toEqual({
      insideBoundingBox: undefined,
      aroundLatLngViaIP: false,
      aroundLatLng: '1,1',
    });
  });

  it('does nothing to aroundLatLng if no defaultPosition parameter is passed', () => {
    const widget = algoliaPlacesWidget({});

    const beforeConfiguration = {
      aroundLatLng: '12,14',
    };

    const afterConfiguration = widget.getConfiguration(beforeConfiguration);

    expect(afterConfiguration).toEqual({});
  });

  it('configures aroundLatLng on change event', () => {
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget(defaultOptions);

    widget.init({ helper });

    const eventName = places.__instance.on.mock.calls[0][0];
    const eventListener = places.__instance.on.mock.calls[0][1];

    expect(eventName).toEqual('change');

    eventListener({ suggestion: { latlng: { lat: '123', lng: '456' } } });

    expect(helper.search).toHaveBeenCalled();
    expect(helper.getState()).toMatchObject({
      insideBoundingBox: undefined,
      aroundLatLng: '123,456',
    });
  });

  it('removes aroundLatLngViaIP on change event', () => {
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget(defaultOptions);

    helper.setQueryParameter('aroundLatLngViaIP', true);

    widget.init({ helper });

    const eventName = places.__instance.on.mock.calls[0][0];
    const eventListener = places.__instance.on.mock.calls[0][1];

    expect(eventName).toEqual('change');

    eventListener({ suggestion: { latlng: { lat: '123', lng: '456' } } });

    expect(helper.search).toHaveBeenCalled();
    expect(helper.getState()).toMatchObject({
      insideBoundingBox: undefined,
      aroundLatLng: '123,456',
      aroundLatLngViaIP: false,
    });
  });

  it('configures aroundLatLng with a default position on clear event', () => {
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget({ defaultPosition: [2, 2] });

    helper.setQueryParameter('aroundLatLngViaIP', true);

    widget.init({ helper });

    const eventName = places.__instance.on.mock.calls[1][0];
    const eventListener = places.__instance.on.mock.calls[1][1];

    expect(eventName).toEqual('clear');

    eventListener();

    expect(helper.search).toHaveBeenCalled();
    expect(helper.getState()).toMatchObject({
      insideBoundingBox: undefined,
      aroundLatLngViaIP: false,
      aroundLatLng: '2,2',
    });
  });

  it('restores aroundLatLngViaIP without a default position on clear event', () => {
    const client = createFakeClient();
    const helper = createFakekHelper(client);
    const widget = algoliaPlacesWidget();

    helper.setQueryParameter('aroundLatLngViaIP', true);

    widget.init({ helper });

    const changeEventName = places.__instance.on.mock.calls[0][0];
    const changeEventListener = places.__instance.on.mock.calls[0][1];

    expect(changeEventName).toEqual('change');

    changeEventListener({
      suggestion: {
        latlng: {
          lat: '123',
          lng: '456',
        },
      },
    });

    expect(helper.search).toHaveBeenCalled();
    expect(helper.getState()).toMatchObject({
      insideBoundingBox: undefined,
      aroundLatLng: '123,456',
      aroundLatLngViaIP: false,
    });

    const clearEventName = places.__instance.on.mock.calls[1][0];
    const clearEventListener = places.__instance.on.mock.calls[1][1];

    expect(clearEventName).toEqual('clear');

    clearEventListener();

    expect(helper.search).toHaveBeenCalled();
    expect(helper.getState()).toMatchObject({
      insideBoundingBox: undefined,
      aroundLatLng: undefined,
      aroundLatLngViaIP: true,
    });
  });

  describe('routing', () => {
    const getInitializedWidget = (
      widgetOptions = { defaultPosition: [2, 2] }
    ) => {
      const client = createFakeClient();
      const helper = createFakekHelper(client);
      const widget = algoliaPlacesWidget(widgetOptions);

      widget.init({ helper, state: helper.state });

      return [widget, helper];
    };

    it('restores aroundLatLngViaIP on clear event', () => {
      const client = createFakeClient();
      const helper = createFakekHelper(client);
      const widget = algoliaPlacesWidget();

      // Simulate the fact that a widget set `aroundLatLngViaIP` from the URLSync
      const searchParametersBefore = SearchParameters.make({
        aroundLatLngViaIP: true,
      });

      const uiState = {
        places: {
          position: '123,123',
          query: 'Paris',
        },
      };

      widget.getWidgetSearchParameters(searchParametersBefore, {
        uiState,
      });

      widget.init({ helper });

      const clearEventListener = places.__instance.on.mock.calls[1][1];

      clearEventListener();

      expect(helper.search).toHaveBeenCalled();
      expect(helper.getState()).toMatchObject({
        insideBoundingBox: undefined,
        aroundLatLng: undefined,
        aroundLatLngViaIP: true,
      });
    });

    describe('getWidgetState', () => {
      test('should give back the object unmodified if the default value is selected', () => {
        const [widget, helper] = getInitializedWidget();
        const uiStateBefore = {};
        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });

        expect(uiStateAfter).toMatchObject(uiStateBefore);
      });

      test('should give update uiState with position', () => {
        const [widget, helper] = getInitializedWidget();

        const eventName = places.__instance.on.mock.calls[0][0];
        const eventListener = places.__instance.on.mock.calls[0][1];

        expect(eventName).toEqual('change');

        eventListener({
          suggestion: {
            latlng: { lat: '123', lng: '456' },
            value: 'Paris',
          },
        });

        expect(helper.search).toHaveBeenCalled();
        expect(helper.getState()).toMatchObject({
          insideBoundingBox: undefined,
          aroundLatLng: '123,456',
        });

        const uiStateBefore = {};

        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });

        const expectedUiState = {
          places: {
            position: '123,456',
            query: 'Paris',
          },
        };

        expect(uiStateAfter).toMatchObject(expectedUiState);
      });

      test('should reset places uiState on clear', () => {
        const [widget, helper] = getInitializedWidget();

        const eventName = places.__instance.on.mock.calls[1][0];
        const eventListener = places.__instance.on.mock.calls[1][1];

        expect(eventName).toEqual('clear');

        eventListener();

        expect(helper.search).toHaveBeenCalled();
        expect(helper.getState()).toMatchObject({
          insideBoundingBox: undefined,
          aroundLatLng: '2,2',
        });

        const uiStateBefore = {
          places: {
            position: '123,456',
            query: 'Paris',
          },
        };

        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });

        const expectedUiState = {
          places: {
            position: '2,2',
            query: '',
          },
        };

        expect(uiStateAfter).toEqual(expectedUiState);
      });

      test('should reset places uiState on clear with no widgetOptions', () => {
        const [widget, helper] = getInitializedWidget({});

        const eventName = places.__instance.on.mock.calls[1][0];
        const eventListener = places.__instance.on.mock.calls[1][1];

        expect(eventName).toEqual('clear');

        eventListener();

        expect(helper.search).toHaveBeenCalled();
        expect(helper.getState()).toMatchObject({
          insideBoundingBox: undefined,
          aroundLatLng: undefined,
        });

        const uiStateBefore = {
          places: {
            position: '123,456',
            query: 'Paris',
          },
        };

        const uiStateAfter = widget.getWidgetState(uiStateBefore, {
          searchParameters: helper.state,
          helper,
        });

        const expectedUiState = {};

        expect(uiStateAfter).toEqual(expectedUiState);
      });
    });

    describe('getWidgetSearchParameters', () => {
      test('should return the same SP if there are no refinements in the UI state', () => {
        const [widget, helper] = getInitializedWidget();
        // The user presses back (browser), and the URL contains no parameters
        const uiState = {};
        // The current state is empty (and page is set to 0 by default)
        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        // Applying the same values should not return a new object
        expect(searchParametersAfter).toBe(searchParametersBefore);
      });

      test('should not care about the default value even if no value is in the UI State', () => {
        const [widget, helper] = getInitializedWidget();
        // The user presses back (browser), and the URL contains no parameters
        const uiState = {};

        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        expect(searchParametersAfter).toMatchSnapshot();
        expect(searchParametersAfter.aroundLatLng).toBe(undefined);
        expect(searchParametersAfter.insideBoundingBox).toBe(undefined);
      });

      test('should reset aroundLatLng if no defaultOptions and no value is in the UI State', () => {
        const [widget, helper] = getInitializedWidget({});
        // The user presses back (browser), and the URL contains no parameters
        const uiState = {};
        // The current state is set to page 4

        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        // Applying an empty state, should force back to page 0
        expect(searchParametersAfter).toMatchSnapshot();
        expect(searchParametersAfter.aroundLatLng).toBe(undefined);
        expect(searchParametersAfter.insideBoundingBox).toBe(undefined);
      });

      test('should add the refinements according to the UI state provided', () => {
        const [widget, helper] = getInitializedWidget();
        // The user presses back (browser), and the URL contains some parameters
        const uiState = {
          places: {
            position: '123,123',
            query: 'Paris',
          },
        };

        const searchParametersBefore = SearchParameters.make(helper.state);
        const searchParametersAfter = widget.getWidgetSearchParameters(
          searchParametersBefore,
          { uiState }
        );
        // Applying a state with new parameters should apply them on the search
        expect(searchParametersAfter).toMatchSnapshot();
        expect(searchParametersAfter.insideBoundingBox).toBe(undefined);
        expect(searchParametersAfter.aroundLatLng).toBe('123,123');
      });

      test('should close the dropdown', () => {
        const [widget, helper] = getInitializedWidget();
        // The user presses back (browser), and the URL contains some parameters
        const uiState = {
          places: {
            position: '123,123',
            query: 'Paris',
          },
        };

        const searchParametersBefore = SearchParameters.make(helper.state);
        widget.getWidgetSearchParameters(searchParametersBefore, { uiState });
        // Applying a state with new parameters should apply them on the search
        expect(places.__instance.close).toHaveBeenCalled();
      });
    });
  });
});
