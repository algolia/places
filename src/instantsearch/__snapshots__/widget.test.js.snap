// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instantsearch widget routing getWidgetSearchParameters should add the refinements according to the UI state provided 1`] = `
SearchParameters {
  "advancedSyntax": undefined,
  "allowTyposOnNumericTokens": undefined,
  "analytics": undefined,
  "analyticsTags": undefined,
  "aroundLatLng": "123,123",
  "aroundLatLngViaIP": false,
  "aroundPrecision": undefined,
  "aroundRadius": undefined,
  "attributesToHighlight": undefined,
  "attributesToRetrieve": undefined,
  "attributesToSnippet": undefined,
  "disableExactOnAttributes": undefined,
  "disjunctiveFacets": Array [],
  "disjunctiveFacetsRefinements": Object {},
  "distinct": undefined,
  "enableExactOnSingleWordQuery": undefined,
  "facets": Array [],
  "facetsExcludes": Object {},
  "facetsRefinements": Object {},
  "getRankingInfo": undefined,
  "hierarchicalFacets": Array [],
  "hierarchicalFacetsRefinements": Object {},
  "highlightPostTag": undefined,
  "highlightPreTag": undefined,
  "hitsPerPage": undefined,
  "ignorePlurals": undefined,
  "index": "",
  "insideBoundingBox": undefined,
  "insidePolygon": undefined,
  "length": undefined,
  "maxValuesPerFacet": undefined,
  "minProximity": undefined,
  "minWordSizefor1Typo": undefined,
  "minWordSizefor2Typos": undefined,
  "minimumAroundRadius": undefined,
  "numericFilters": undefined,
  "numericRefinements": Object {},
  "offset": undefined,
  "optionalFacetFilters": undefined,
  "optionalTagFilters": undefined,
  "optionalWords": undefined,
  "page": 0,
  "query": "",
  "queryType": undefined,
  "removeWordsIfNoResults": undefined,
  "replaceSynonymsInHighlight": undefined,
  "restrictSearchableAttributes": undefined,
  "snippetEllipsisText": undefined,
  "synonyms": undefined,
  "tagFilters": undefined,
  "tagRefinements": Array [],
  "typoTolerance": undefined,
}
`;

exports[`instantsearch widget routing getWidgetSearchParameters should not care about the default value even if no value is in the UI State 1`] = `
SearchParameters {
  "advancedSyntax": undefined,
  "allowTyposOnNumericTokens": undefined,
  "analytics": undefined,
  "analyticsTags": undefined,
  "aroundLatLng": undefined,
  "aroundLatLngViaIP": undefined,
  "aroundPrecision": undefined,
  "aroundRadius": undefined,
  "attributesToHighlight": undefined,
  "attributesToRetrieve": undefined,
  "attributesToSnippet": undefined,
  "disableExactOnAttributes": undefined,
  "disjunctiveFacets": Array [],
  "disjunctiveFacetsRefinements": Object {},
  "distinct": undefined,
  "enableExactOnSingleWordQuery": undefined,
  "facets": Array [],
  "facetsExcludes": Object {},
  "facetsRefinements": Object {},
  "getRankingInfo": undefined,
  "hierarchicalFacets": Array [],
  "hierarchicalFacetsRefinements": Object {},
  "highlightPostTag": undefined,
  "highlightPreTag": undefined,
  "hitsPerPage": undefined,
  "ignorePlurals": undefined,
  "index": "",
  "insideBoundingBox": undefined,
  "insidePolygon": undefined,
  "length": undefined,
  "maxValuesPerFacet": undefined,
  "minProximity": undefined,
  "minWordSizefor1Typo": undefined,
  "minWordSizefor2Typos": undefined,
  "minimumAroundRadius": undefined,
  "numericFilters": undefined,
  "numericRefinements": Object {},
  "offset": undefined,
  "optionalFacetFilters": undefined,
  "optionalTagFilters": undefined,
  "optionalWords": undefined,
  "page": 0,
  "query": "",
  "queryType": undefined,
  "removeWordsIfNoResults": undefined,
  "replaceSynonymsInHighlight": undefined,
  "restrictSearchableAttributes": undefined,
  "snippetEllipsisText": undefined,
  "synonyms": undefined,
  "tagFilters": undefined,
  "tagRefinements": Array [],
  "typoTolerance": undefined,
}
`;

exports[`instantsearch widget routing getWidgetSearchParameters should reset aroundLatLng if no defaultOptions and no value is in the UI State 1`] = `
SearchParameters {
  "advancedSyntax": undefined,
  "allowTyposOnNumericTokens": undefined,
  "analytics": undefined,
  "analyticsTags": undefined,
  "aroundLatLng": undefined,
  "aroundLatLngViaIP": undefined,
  "aroundPrecision": undefined,
  "aroundRadius": undefined,
  "attributesToHighlight": undefined,
  "attributesToRetrieve": undefined,
  "attributesToSnippet": undefined,
  "disableExactOnAttributes": undefined,
  "disjunctiveFacets": Array [],
  "disjunctiveFacetsRefinements": Object {},
  "distinct": undefined,
  "enableExactOnSingleWordQuery": undefined,
  "facets": Array [],
  "facetsExcludes": Object {},
  "facetsRefinements": Object {},
  "getRankingInfo": undefined,
  "hierarchicalFacets": Array [],
  "hierarchicalFacetsRefinements": Object {},
  "highlightPostTag": undefined,
  "highlightPreTag": undefined,
  "hitsPerPage": undefined,
  "ignorePlurals": undefined,
  "index": "",
  "insideBoundingBox": undefined,
  "insidePolygon": undefined,
  "length": undefined,
  "maxValuesPerFacet": undefined,
  "minProximity": undefined,
  "minWordSizefor1Typo": undefined,
  "minWordSizefor2Typos": undefined,
  "minimumAroundRadius": undefined,
  "numericFilters": undefined,
  "numericRefinements": Object {},
  "offset": undefined,
  "optionalFacetFilters": undefined,
  "optionalTagFilters": undefined,
  "optionalWords": undefined,
  "page": 0,
  "query": "",
  "queryType": undefined,
  "removeWordsIfNoResults": undefined,
  "replaceSynonymsInHighlight": undefined,
  "restrictSearchableAttributes": undefined,
  "snippetEllipsisText": undefined,
  "synonyms": undefined,
  "tagFilters": undefined,
  "tagRefinements": Array [],
  "typoTolerance": undefined,
}
`;
